cloudant-http/src/main/java/com/cloudant/http/internal/interceptors/CookieInterceptorBase.java

```diff
@@ -0,0 +1,269 @@
+        try {
+            if (shouldAttemptCookieRequest.get()) {
+
+                HttpURLConnection connection = context.connection.getConnection();
+                int statusCode = connection.getResponseCode();
+                if (statusCode == HttpURLConnection.HTTP_UNAUTHORIZED) {
+                    boolean success = requestCookie(context);
+                    if (success) {
+                        context.replayRequest = true;
+                    } else {
+                        context.replayRequest = false; // Don't replay
+                        shouldAttemptCookieRequest.set(false); // Set the flag to stop trying
+                    }
+                } else {
+                    // Store any cookies provided on the response
+                    storeCookiesFromResponse(connection);
+                }
+            }
+        } catch (IOException e) {
+            logger.log(Level.SEVERE, "Error reading response code or body from request", e);
+        }
+        return context;
+
+    }
+
+    // helper for requestCookie
+    interface OnExecuteCallable {
+        boolean call(HttpConnection connection) throws IOException;
+    }
+
+    protected boolean requestCookie(HttpConnectionInterceptorContext context,
+                                    URL url,
+                                    byte[] payload,
+                                    String mimeType,
+                                    String accept,
+                                    OnExecuteCallable onResponseOk) {
+
+        try {
+            HttpConnection conn = Http.POST(url, mimeType);
+            conn.requestProperties.put("accept", accept);
+            conn.setRequestBody(payload);
+
+            //when we request the session we need all interceptors except this one
+
+            conn.requestInterceptors.addAll(context.connection.requestInterceptors);
+            conn.requestInterceptors.remove(this);
+            conn.responseInterceptors.addAll(context.connection.responseInterceptors);
+            conn.responseInterceptors.remove(this);
+
+            HttpConnection connection = conn.execute();
+            int responseCode = connection.getConnection().getResponseCode();
+
+            if (responseCode / 100 == 2) {
+                return onResponseOk.call(connection);
+            } else {
+                InputStream errorStream = connection.getConnection().getErrorStream();
+                try {
+                    if (errorStream != null) {
+                        // Consume the error stream to avoid leaking connections
+                        String error = IOUtils.toString(errorStream, "UTF-8");
+                        // Log the error stream content
+                        logger.fine(error);
+                    }
+                } finally {
+                    IOUtils.closeQuietly(errorStream);
+                }
+                // TODO these messages should be customised so we can tell if bearer or session token failed
```

Comment:
Let's resolve this TODO by adding the URL to the message which will make it clear which endpoint caused a problem.